src/shared/misc.c

```diff
@@ -43,6 +43,106 @@
 
 extern char **environ;
 
+
+static char *deptree_cache_path = NULL;
+static char *deptree_skewed_path = NULL;
+static char *krunlevel_path = NULL;
+static char *starting_path = NULL;
+static char *stopping_path = NULL;
+
+static char *svcdir_starting = NULL;
+static char *svcdir_inactive = NULL;
+static char *svcdir_started = NULL;
+static char *svcdir_coldplugged = NULL;
+
+char *
+rc_deptree_cache_path_get() {
+	if (!deptree_cache_path) {
+		deptree_cache_path = xmalloc(sizeof(char) * PATH_MAX);
+		snprintf(deptree_cache_path, PATH_MAX, "%s/%s",
+				rc_svcdir_get(), RC_DEPTREE_CACHE);
+	}
+	return deptree_cache_path;
+}
+
+char *
+rc_deptree_skewed_path_get() {
+	if (!deptree_skewed_path) {
+		deptree_skewed_path = xmalloc(sizeof(char) * PATH_MAX);
+		snprintf(deptree_skewed_path, PATH_MAX, "%s/%s",
+				rc_svcdir_get(), RC_DEPTREE_SKEWED);
+	}
+	return deptree_skewed_path;
+}
+
+char *
+rc_krunlevel_path_get() {
+	if (!krunlevel_path) {
+		krunlevel_path = xmalloc(sizeof(char) * PATH_MAX);
```

Comment:
Hi @anna-cli. One major problem I see is that memory is allocated each time the function is called, but never freed. Or did I miss something?

Reply:
> Hi [@anna-cli](https://github.com/anna-cli). One major problem I see is that memory is allocated each time the function is called, but never freed. Or did I miss something?

Hello [@andy5995](https://github.com/andy5995)! It shouldn't allocate more than once per function (since the
pointers are global, and there's the check before allocation). But you
made me realise that is probably better to just declare the string as a
static char array inside the function!